timemachine/cpp/src/nonbonded_pairs.cu

```diff
@@ -0,0 +1,143 @@
+#include "gpu_utils.cuh"
+#include "k_nonbonded_pairs.cuh"
+#include "nonbonded_pairs.hpp"
+#include <stdexcept>
+#include <vector>
+
+namespace timemachine {
+
+template <typename RealType, bool Interpolated>
+NonbondedPairs<RealType, Interpolated>::NonbondedPairs(
+    const std::vector<int> &pair_idxs,          // [M, 2]
+    const std::vector<double> &scales,          // [M, 2]
+    const std::vector<int> &lambda_plane_idxs,  // [N]
+    const std::vector<int> &lambda_offset_idxs, // [N]
+    const double beta,
+    const double cutoff,
+    const std::string &kernel_src)
+    : N_(lambda_offset_idxs.size()), M_(pair_idxs.size() / 2), beta_(beta), cutoff_(cutoff),
+      compute_w_coords_instance_(kernel_cache_.program(kernel_src.c_str()).kernel("k_compute_w_coords").instantiate()),
+      compute_permute_interpolated_(
+          kernel_cache_.program(kernel_src.c_str()).kernel("k_permute_interpolated").instantiate()) {
+
+    if (pair_idxs.size() % 2 != 0) {
+        throw std::runtime_error("pair_idxs.size() must be exactly 2*M");
+    }
+
+    for (int i = 0; i < M_; i++) {
+        auto src = pair_idxs[i * 2 + 0];
+        auto dst = pair_idxs[i * 2 + 1];
+        if (src == dst) {
+            throw std::runtime_error("illegal pair with src == dst");
+        }
+    }
+
+    if (scales.size() / 2 != M_) {
+        throw std::runtime_error("bad scales size!");
+    }
+
+    gpuErrchk(cudaMalloc(&d_pair_idxs_, M_ * 2 * sizeof(*d_pair_idxs_)));
+    gpuErrchk(cudaMemcpy(d_pair_idxs_, &pair_idxs[0], M_ * 2 * sizeof(*d_pair_idxs_), cudaMemcpyHostToDevice));
+
+    gpuErrchk(cudaMalloc(&d_lambda_plane_idxs_, N_ * sizeof(*d_lambda_plane_idxs_)));
+    gpuErrchk(cudaMemcpy(
+        d_lambda_plane_idxs_, &lambda_plane_idxs[0], N_ * sizeof(*d_lambda_plane_idxs_), cudaMemcpyHostToDevice));
+
+    gpuErrchk(cudaMalloc(&d_lambda_offset_idxs_, N_ * sizeof(*d_lambda_offset_idxs_)));
+    gpuErrchk(cudaMemcpy(
+        d_lambda_offset_idxs_, &lambda_offset_idxs[0], N_ * sizeof(*d_lambda_offset_idxs_), cudaMemcpyHostToDevice));
+
+    gpuErrchk(cudaMalloc(&d_w_, N_ * sizeof(*d_w_)));
+    gpuErrchk(cudaMalloc(&d_dw_dl_, N_ * sizeof(*d_dw_dl_)));
+    gpuErrchk(cudaMalloc(&d_du_dp_buffer_, N_ * 3 * sizeof(*d_du_dp_buffer_)));
+
+    gpuErrchk(cudaMalloc(&d_dp_dl_, N_ * 3 * sizeof(*d_dp_dl_)));
+
+    gpuErrchk(cudaMalloc(&d_scales_, M_ * 2 * sizeof(*d_scales_)));
+    gpuErrchk(cudaMemcpy(d_scales_, &scales[0], M_ * 2 * sizeof(*d_scales_), cudaMemcpyHostToDevice));
+
+    // construct identity permutation
+    if (Interpolated) {
+        gpuErrchk(cudaMalloc(&d_perm_, N_ * sizeof(*d_perm_)));
+        for (int i = 0; i < N_; i++) {
```

Comment:
Can you set GPU memory this way? Pretty sure this needs to be done through code like this: https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/nonbonded.cu#L223-L229

Reply:
makes sense - I guess we could also allocate and initialize on the host and copy? (since this is only happening at construction, I don't think it'd be a bottleneck?)

Reply:
Allocating on the host would also work

Reply:
Should be fixed to allocate on host and copy as of e849535.